### PR TITLE
Set up a GameState abstraction

### LINKED ISSUE #31: Set up a GameState abstraction in `app/lib`
Original discussion [here](https://github.com/snappy-cobra/WiiDevEnv/pull/22#discussion_r1048488264).

A nice abstraction to make working on the game (and testing it) easy is to essentially have the following:

```rust
pub struct GameState {
  world: hecs::World,
  // other global game data like score, toggle between modes, etc.
}

pub struct Changes {
  pub controls: Controls,
  pub delta_time_ms: u64,
  // Other things we expect the game loop/Wii hardware to provide each frame
}

pub struct Controls {
  home_button_down: bool,
  one_button_down: bool,
  // Etc. Any current controller state we are interested in including where the wiimote(s) point.
}

impl GameState {
  /// Set up anything required for a new game
  pub fn new() -> GameState {
    // ...
  }

   /// Performs game logic of one iteration of the game loop.
  /// Before calling this, the caller should have filled `Changes` with the required values.
  /// 
  /// If the returned bool is `true`, the game loop should continue. 
  /// When `false` is returned, the game should quit.
  pub fn update(&mut GameState, &Changes) -> bool {
    // ...
  }
}
```

This is what is defined inside the `app/lib` subcrate. `Changes` is filled from the outside, `GameState` is only read (like in the renderer) but never written to from the outside.

This way we can structure the application akin what is known as 'hexagonal archtecture' in OOP land or 'imperative shell, pure core' in FP land. (Up to a point; even nicer (more testable) would be to have `update` be a function returninga new GameState rather than mutating one in place but I believe that does not combine well with ECS).

### PR DESCRIPTION
**Pull request by Qqwy**
Fixes #31 

- New GameState abstraction in lib and related structs.
- Moves a lot of functionality from `game.rs` to `lib/src/lib.rs`.
  - Note that this can be refactored a lot further in the future, but I recommend doing that after this PR is merged.
- Introduce InputManager in `main.rs`
- Simplify the main function now that the new abstractions exist.

### CODE REVIEW

File: app/lib/src/lib.rs [hunk 1]
```diff
@@ -8,12 +8,183 @@ extern crate alloc;
...
+        system_bounce_bounds(&mut self.world);
+        system_integrate_motion(&mut self.world);
+
+        true
+    }
+}
+
+
+
+
```
Comment (LucvandenBrand) [inline app/lib/src/lib.rs]: Is clippy still working? I'd expect these whitelines to be caught by it :thinking:

Comment (Qqwy) [reply app/lib/src/lib.rs]: Strange. Seems like there was a race condition between the formatter and my editor.

Resolved by running the formatter manually.

File: app/src/main.rs [hunk 2]
```diff
@@ -45,43 +44,36 @@ fn main(_argc: isize, _argv: *const *const u8) -> isize {
...
     let renderer = Renderer::new();
+
     while KEEP_RUNNING.load(Ordering::SeqCst) {
-        Input::update(ControllerType::Wii);
-        if wii_mote.is_button_down(Button::Home) {
+        let controls = input_manager.update();
+
+        let changes = Changes {
+            controls,
+            delta_time_ms: 100,
```
Comment (LucvandenBrand) [inline app/src/main.rs:65]: I assume this is due to us not being able to get the real delta time?

File: app/src/main.rs [hunk 3]
```diff
@@ -45,43 +44,36 @@ fn main(_argc: isize, _argv: *const *const u8) -> isize {
...
-        }
-        system_bounce_bounds(&mut world, &mut all_query);
-        system_integrate_motion(&mut world, &mut all_query);
 
-        renderer.render_world(&world);
+        renderer.render_world(&game_state.world);
     }
-    shutdown();
+    println!("Test me");
+    // shutdown();
```
Comment (LucvandenBrand) [inline app/src/main.rs]: Feel free to remove the commented-out code, not only this line as well :)

File: app/src/main.rs [hunk 4]
```diff
@@ -104,8 +96,39 @@ extern "C" fn power_callback() {
...
-    unsafe {
-        STM_ShutdownToStandby();
+    // unsafe {
+    //     STM_ShutdownToStandby();
+    // }
+    unsafe { exit(0) };
+    // core::unreachable!()
+}
+
+struct InputManager {
```
Comment (LucvandenBrand) [inline app/src/main.rs]: Again, sorry if I'm nitpicky and I feel like you've explained this to me before, but... :see_no_evil: 
Wouldn't it be neater to separate these things into separate modules like `input.rs`?

File: app/src/main.rs (hunk 2 above)
Comment (Qqwy) [reply app/src/main.rs:65]: Correct! I included it as an example of other data we want to pass along between the outside world and the 'pure' world inside.

File: app/src/main.rs (hunk 3 above)
Comment (Qqwy) [reply app/src/main.rs]: This garbage should not have been committed by me; it was unfinished testing code from another branch 😅

File: app/src/main.rs (hunk 4 above)
Comment (Qqwy) [reply app/src/main.rs]: Good idea!

File: app/lib/src/lib.rs [hunk 5]
```diff
@@ -8,12 +8,175 @@ extern crate alloc;
...
+    pub controls: Controls,
+    pub delta_time_ms: u64,
+}
+
+/// Represents the state of one or multiple Wii controllers w.r.t. the game
+///
+/// Its fields are public so they can be filled in from outside the library
+/// (before `GameState::update` is called)
+#[derive(Debug, Default)]
+pub struct Controls {
```
Comment (LucvandenBrand) [inline app/lib/src/lib.rs:45]: I'd put ` Controls`  in the `input` module, what do you think?

**Review by LucvandenBrand — APPROVED** [review]
Just one minor comment, feel free to do something with it or not, but I'm very happy with this in general :)

**Comment by Qqwy** [comment]
Thanks! Feel free to merge it because I'll only be back at my computer tomorrow afternoon.

About the location of the Controls struct: This is inside lib rather than in the outer project's input module because otherwise the lib project cannot use it.

The idea to put it and the Changes struct in lib as a whole (rather than mutating the game state from the outside altogether) is because this way we can write nice and simple tests of the kind "given a mario entity standing on a platform, when someone presses A, then mario starts a jump".

On December 30, 2022 10:22:55 AM UTC, Luc van den Brand ***@***.***> wrote:
>@LucvandenBrand approved this pull request.
>
>Just one minor comment, feel free to do something with it or not, but I'm very happy with this in general :)
>
>> +/// Represents any changes made by the outside world during a single frame.
>+///
>+/// Its fields are public so they can be filled in from outside the library
>+/// (before `GameState::update` is called)
>+#[derive(Debug, Default)]
>+pub struct Changes {
>+    pub controls: Controls,
>+    pub delta_time_ms: u64,
>+}
>+
>+/// Represents the state of one or multiple Wii controllers w.r.t. the game
>+///
>+/// Its fields are public so they can be filled in from outside the library
>+/// (before `GameState::update` is called)
>+#[derive(Debug, Default)]
>+pub struct Controls {
>
>I'd put ` Controls`  in the `input` module, what do you think?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/snappy-cobra/WiiDevEnv/pull/52#pullrequestreview-1232998002
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>
-- 
Met vriendelijke groet / Sincerely, 
~W-M

**Pull request merged**